xarray/backends/api.py

```diff
@@ -542,7 +527,99 @@ def maybe_decode_store(store, lock=False):
+        if engine != 'zarr':
+
+            # if passed an actual file path, augment the token with
+            # the file modification time
+            if isinstance(filename_or_obj, str) and not is_remote_uri(filename_or_obj):
+                mtime = os.path.getmtime(filename_or_obj)
+            else:
+                mtime = None
+            token = tokenize(
+                filename_or_obj,
+                mtime,
+                group,
+                decode_cf,
+                mask_and_scale,
+                decode_times,
+                concat_characters,
+                decode_coords,
+                engine,
+                chunks,
+                drop_variables,
+                use_cftime,
+            )
+            name_prefix = "open_dataset-%s" % token
+            ds2 = ds.chunk(chunks, name_prefix=name_prefix, token=token)
+            ds2._file_obj = ds._file_obj
+
+        else:  # file is zarr!
+
+            # adapted from Dataset.Chunk() and taken from open_zarr
+            if not isinstance(chunks, (int, dict)):
+                if chunks != "auto":
+                    raise ValueError(
+                        "chunks must be an int, dict, 'auto', or None. "
+                        "Instead found %s. " % chunks
+                    )
+            if isinstance(chunks, int):
+                chunks = dict.fromkeys(ds.dims, chunks)
+
+            if isinstance(chunks, tuple) and len(chunks) == len(ds.dims):
+                chunks = dict(zip(ds.dims, chunks))
+
+            def get_chunk(name, var, chunks):
+                chunk_spec = dict(zip(var.dims, var.encoding.get("chunks")))
+
+                # Coordinate labels aren't chunked
+                if var.ndim == 1 and var.dims[0] == name:
+                    return chunk_spec
+
+                if chunks == "auto":
+                    return chunk_spec
+
+                for dim in var.dims:
+                    if dim in chunks:
+                        spec = chunks[dim]
+                        if isinstance(spec, int):
+                            spec = (spec,)
+                        if isinstance(spec, (tuple, list)) and chunk_spec[dim]:
+                            if any(s % chunk_spec[dim] for s in spec):
+                                warnings.warn(
+                                    "Specified Dask chunks %r would "
+                                    "separate Zarr chunk shape %r for "
+                                    "dimension %r. This significantly "
+                                    "degrades performance. Consider "
+                                    "rechunking after loading instead."
+                                    % (chunks[dim], chunk_spec[dim], dim),
+                                    stacklevel=2,
+                                )
+                        chunk_spec[dim] = chunks[dim]
+                return chunk_spec
+
+            def maybe_chunk(name, var, chunks):
```

Comment:
I think I have a slight preference to relocate `maybe_chunk` and `get_chunk` to the `zarr.py` module. Just skimming here but it doesn't seem like they are using local function state.

Reply:
It does make sense, and would make the code cleaner/easier to read and/or debug. You think it should be implemented within ``ZarrStore.open_group``?